### DISCUSSION TITLE
Protect / authenticate SSG pages

### DISCUSSION THREAD

**Discussion by JeroenReumkens** [vercel/next.js#20413] [Help]
I'm currently building a NextJS website that is fully authenticated, but has the exact same content for every user. Since this content is exactly the same for each user, I want to pre render all the HTML pages during the build (SSG) and let Next/Express only show the pages if the user is authenticated. 

**I'm currently doing that as follow:** 
* During build generate all the pages with authenticated content to static HTML
* Serve files via a custom Express server. This checks whether you are authenticated and returns either the HTML or redirects you to /login

**However, this brings up one problem I hope to solve:** 
If the user navigates via client side routing to an authenticated page, the page will be rendered. I could protect this page via a AuthProvider that could also redirect you. But in that case I think the page data / content will still be sent to the client / can be found by searching the JS bundles, right? 

So I'm looking for a way to pre render this authenticated content, but still keeping the content protected. Probably resolving it always via express? 

**Solutions I can think of (and have downsides)**
1. All `_next` directory files are also resolved via express, so I could maybe to a redirect on these? Possible downsides could be: I don't know if next will accept those redirects, and for each page there are multiple `_next/` files requested, meaning I'm doing way more requests to the backend api to verify the user the necessary. 
2. Don't use SSG, just use SSR. As far as I could find online I have a feeling that my `getServerSideProps` method will be called for each user to fetch the data (although the data is the same for each user), resulting in speed loss compared to just serving the static file.

Hoping anyone has a neat solution for this, or someone can tell me what I'm thinking of is stupid 😅

Thanks in advance for any help on this!

**Comment by jozsefm** [comment]
I have asked basically the same question with a slightly different use case with #20185. No real answers yet for me, but some of the suggestions in the comments there might help you or give inspiration. 

I think the general idea of the recommended tools there goes like this:
Create the whole site as if there would be no authentication at all, pregenerate it to static html and simply add a middleware to the custom express server to handle checking the authentication status. Let the request through to the next sever to handle it as it should when everything is OK or simply redirect them to the login page, for which you should add an exception to the auth checking.

**Reply by jozsefm** [reply]
As I was browsing the Discussions I stumbled into #13705 that might help to achieve a similar effect by caching the results of an SSR-ed page. I will evaluate it for my own use case. Might worth a look for you too.

**Reply by JeroenReumkens** [reply]
Thanks a lot for your reply! Somehow I didn’t find your thread when searching for it. 

Seems like we indeed have a quite similar question. Hopefully anyone will stumble across and have the answer for us.

I really can not imagine that we are the only one with the use case of statically generating pages but still wanting to protect them via Express.
